### PR TITLE
Fix multi line paste detection and filtering

### LINKED ISSUE #5821: Terminal doesn't translate LF to CR on paste; it only translates CRLF
Notes from @DHowett-MSFT:

## Line ending conversions by other terminals

Clipboard Contains | xterm sends | wt sends | putty sends | gnome-terminal sends | conhost sends
-|-|-|-|-|-
`\r\n` (CRLF)|`\r` ✔️|`\r` ✔️|`\r` ✔️|`\r` ✔️|`\r` ✔️
`\r` (CR)    |`\r` ✔️|`\r` ✔️|`\r` ✔️|`\r` ✔️|`\r` ✔️
`\n` (LF)    |`\r` ✔️|`\n` ❌|`\n` ❌|`\r` ✔️|`\r` ✔️

I guess that means that both of WT and putty are wrong :smile:

--- original filing below ---

# Environment

```none
Windows 10 10.0.19041.207
Windows Terminal 0.11.1251.0
```

# Steps to reproduce

1. Open Windows Terminal (Windows PowerShell or PowerShell Core).
2. Copy all the text in the attached file [fake_key.txt](https://github.com/microsoft/terminal/files/4602399/fake_key.txt).
3. Paste into Windows Terminal with the mouse.
4. Observe the lines appear in reverse order.

# Expected behavior

Similar behavior to cmd.exe and wsl, in that the pasted lines are in original order and intact.

# Actual behavior

Bamboozled. Lines are presented in reverse order.

# Screenshots

image

image

image

### LINKED ISSUE #8601: Multi-line paste warning doesn't detect \r, only \n
Certain shells will execute commands on \r as well as \n, so we should try to catch \r as well.

This should probably be tackled at the same time as paste filtering. I originally thought that we should detect newlines _after_ filtering, or perhaps _during_ filtering, but would result in us doing a "lot" of work in the case where the user declines the paste.

Filtering: #5821.

### PR DESCRIPTION
**Pull request by PankajBhojwani**
- Detect `\r` when warning about multi line paste
- Translate `\n` to `\r` on paste

## PR Checklist
* [x] Closes #8601
* [x] Closes #5821

## Validation Steps Performed
Manual testing

### CODE REVIEW

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 1]
```diff
@@ -1973,7 +1973,8 @@ namespace winrt::TerminalApp::implementation
                 }
             }
 
-            const bool hasNewLine = std::find(text.cbegin(), text.cend(), L'\n') != text.cend();
+            std::wregex newLine{ LR"((\r|\n))" };
```
Comment (Don-Vito) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1976]: Consider using non-capturing groups, aka `std::regex_constants::nosubs`

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -2025,29 +2025,13 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
-        std::wstring stripped{ wstr };
-
-        std::wstring::size_type pos = 0;
-
-        while ((pos = stripped.find(L"\r\n", pos)) != std::wstring::npos)
-        {
-            stripped.replace(pos, 2, L"\r");
-        }
+        std::wregex multiLines{ LR"(((\r)?)(\n))" };
+        auto res = std::regex_replace(wstr, multiLines, L"\r");
```
Comment (Don-Vito) [inline src/cascadia/TerminalControl/TermControl.cpp:2032]: Same as above

Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/TermControl.cpp]: Thanks for all the suggestions! No longer using regex now :)

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -1973,9 +1973,9 @@ namespace winrt::TerminalApp::implementation
                 }
             }
 
-            std::wregex newLine{ LR"((\r|\n))" };
-            const bool hasNewLine = std::regex_search(text.c_str(), newLine);
-            const bool warnMultiLine = hasNewLine && _settings.GlobalSettings().WarnAboutMultiLinePaste();
+            const auto hasNewLineLambda = [&](auto c) { return c == L'\n' || c == L'\r'; };
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:1976]: no need for capture here :smile:

```suggestion
            const auto hasNewLineLambda = [](auto c) { return c == L'\n' || c == L'\r'; };
```

nit: prefer name `isNewLineLambda` (you're finding it "if it is a newline")

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -2028,10 +2028,36 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+            }
+            else
+            {
+                // We found a lone '\n', replace it with '\r'
+                stripped.replace(pos, 1, L"\r");
+            }
+        }
 
-        _connection.WriteInput(res);
+        _connection.WriteInput(stripped);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp]: nit: if you're interested in a minor optimization... this does do one more string copy than necessary.

There might be a way you can do it with `stripped` being an empty string w/ a reservation, and only pass over the source string twice (instead of 3 times)

There's _also_ a way for you to do it without copying in the case where there is no newline

Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/TermControl.cpp]: fixed! thanks for the suggestion

**Review by DHowett — APPROVED** [review]
one shot, one target eliminated. excellent work

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 5]
```diff
@@ -2061,7 +2061,7 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
             begin = pos;
         }
 
-        if (begin == 0 && pos == std::wstring::npos)
+        if (begin == 0)
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2064]: I'm not certain I agree with that. If you pass in `hello\n`:

```
stripped = hello\r
wstr     = hello\n
begin    = 0
pos      = 5
```

we will erroneously send wstr (`\n`) instead of stripped (`\r`)

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 5 above)
Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/TermControl.cpp:2064]: Wouldn't `begin` be equal to 5 because of line 2061?

Also since this is after the while loop, `pos` is guaranteed to be `npos` right? Otherwise we would not have exited the loop

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.cpp:2064]: oh great catch

Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.cpp:2064]: does it deserve a comment explaining how we know we're safe here?

Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/TermControl.cpp:2064]: Okay will do!

Comment (PankajBhojwani) [reply src/cascadia/TerminalControl/TermControl.cpp:2068]: Done

**Review by DHowett — APPROVED** [review]

**Review by zadjii-msft — APPROVED** [review]
Okay that seems fine. I'd feel better with a test for this, since it's a little hard to follow, but there's not a good way of doing TermControl unittests at the moment, so I just stuck this on #7001 for now.
